vm/concurrent_hash.go

```diff
@@ -0,0 +1,168 @@
+package vm
+
+import (
+	"sync"
+
+	"github.com/goby-lang/goby/vm/classes"
+	"github.com/goby-lang/goby/vm/errors"
+)
+
+// Pseudo-constant definition of the forwarded methods, mapped to a boolean representing the
+// requirement for a write lock (true) or read lock (false)
+var ConcurrentHashMethodsForwardingTable = map[string]bool{
+	"[]": false,
+	"[]=": true,
+	"any?": false,
+	"clear": true,
+	"default": false,
+	"default=": true,
+	"delete": true,
+	"delete_if": true,
+	"dig": false,
+	"each": false,
+	"each_key": false,
+	"each_value": false,
+	"empty?": false,
+	"eql?": false,
+	"fetch": false,
+	"has_key?": false,
+	"has_value?": false,
+	"keys": false,
+	"length": false,
+	"map_values": false,
+	"merge": false,
+	"select": false,
+	"sorted_keys": false,
+	"to_a": false,
+	"to_json": false,
+	"to_s": false,
+	"transform_values": false,
+	"values": false,
+	"values_at": false,
+}
+
+// ConcurrentHashObject is a thread-safe Hash.
+//
+// The current design is the simplest possible, via a R/W mutex; if/when performance will be a
+// concern, it's trivial to write more sophisticated implementations, since the logic is entirely
+// encapsulated.
+//
+// The class is current a subclass of HashObject internally, but an Object direct subclass at
+// user level.
+type ConcurrentHashObject struct {
+	*baseObj
+	InternalHash *HashObject
+
+	sync.RWMutex
```

Comment:
You should not embed the original HashObject with lock. This performs poorly than using `sync.Map` as data storage directly. See this:
![](https://cdn-images-1.medium.com/max/1600/1*8sCf9LsFhVcZLMw45kn0nQ.png)

[Reference](https://medium.com/@deckarep/the-new-kid-in-town-gos-sync-map-de24a6bf7c2c)

Reply:
I know that article :wink:; it's import to have clear that it's not meant to be used as a reference for the general speed of `sync.Map` (it's actually explicitly written both in the article, and in the sync.Map documentation).

`sync.Map` is designed with a specific case in mind (essentially, high volume reads with sparse writes), and the article tests that case only, in ideal conditions. Under other circumstances (for example, certain caching patterns), `sync.Map` will have [comparable or worse performance](https://golang.org/pkg/sync/#Map).

With this in mind, since the Goby implementation targets no specific use case, until there is no real-world usage, both are equally valid.  

Probably a good idea would be to provide Goby mutex APIs, so that if users want to implement a lock-based hash, they can.

Reply:
@saveriomiroddi Sorry for not completely understanding the reference and provided wrong suggestion 😢. I think I should take a time to have a deeper understanding on concurrent programming. Do you think [this](https://stackoverflow.com/questions/19148809/how-to-use-rwmutex-in-golang#answer-19168242) is a clear and correct explanation of Go's `sync.Mutex` and `sync.RWMutex`?

Reply:
> Sorry for not completely understanding the reference and provided wrong suggestion

No problem :smile: 

Actually, I think that by providing synchronization APIs (#477), the new implementation will give users (programmers) flexibility for all the use cases:

- ConcurrentHash, based on sync.Map, will provide optimal performance for the mentioned use case
- regular Hash + synchronization APIs, for all the use cases where ConcurrentHash doesn't perform

With the previous implementation (ConcurrentHash based on RWLock), there were choice for users (programmers).

I think we can follow up with the discussion on #477.

Very good! :+1: 

> Do you think this is a clear and correct explanation of Go's sync.Mutex and sync.RWMutex

I think so, yes!